File: modules/consultorias.py
"""Módulo de Consultorias (Private, Black, Retenção)."""
import pandas as pd
import streamlit as st

from core import db, ui
from core.slack import enviar_slack

TIPOS = ["Private", "Black", "Retenção"]
STATUS = ["Solicitada", "Aprovada", "Agendada", "Realizada", "Cancelada"]


def render(user):
    ui.titulo("Consultorias", "Controle de consultorias Private, Black e Retenção")
    tab_lista, tab_nova = st.tabs(["📋 Consultorias", "➕ Nova consultoria"])

    with tab_nova:
        with st.form("nova_cons", clear_on_submit=True):
            c1, c2, c3 = st.columns(3)
            aluno = c1.text_input("Nome do aluno *")
            tipo = c2.selectbox("Tipo", TIPOS)
            solicitante = c3.text_input("Solicitante", value=user["nome"])
            c1, c2, c3 = st.columns(3)
            data_sol = c1.date_input("Data solicitada")
            data_ag = c2.date_input("Data agendada")
            responsavel = c3.text_input("Responsável interno")
            c1, c2 = st.columns(2)
            status = c1.selectbox("Status", STATUS)
            zoom = c2.text_input("Link Zoom")
            obs = st.text_area("Observações")
            if st.form_submit_button("Criar consultoria", type="primary"):
                if not aluno:
                    st.error("Informe o nome do aluno.")
                else:
                    db.execute(
                        "INSERT INTO consultorias (aluno,tipo,solicitante,data_solicitada,"
                        "data_agendada,responsavel,status,link_zoom,observacoes,criado_em)"
                        " VALUES (?,?,?,?,?,?,?,?,?,?)",
                        (aluno, tipo, solicitante, str(data_sol), str(data_ag), responsavel,
                         status, zoom, obs, db.now()))
                    db.log_action(user["email"], "consultoria_criada", f"{tipo} · {aluno}")
                    enviar_slack(f":telephone_receiver: *Nova consultoria {tipo}* — {aluno} "
                                 f"(solicitante {solicitante}, resp. {responsavel})", "consultoria_nova")
                    st.success("Consultoria criada — alerta enviado ao Slack.")

    with tab_lista:
        c1, c2, c3 = st.columns(3)
        f_tipo = c1.multiselect("Tipo", TIPOS)
        f_status = c2.multiselect("Status", STATUS)
        f_busca = c3.text_input("Buscar aluno")
        rows = db.query("SELECT * FROM consultorias ORDER BY data_agendada DESC")
        data = []
        for r in rows:
            if f_tipo and r["tipo"] not in f_tipo:
                continue
            if f_status and r["status"] not in f_status:
                continue
            if f_busca and f_busca.lower() not in (r["aluno"] or "").lower():
                continue
            data.append({
                "ID": r["id"], "Aluno": r["aluno"], "Tipo": r["tipo"],
                "Solicitante": r["solicitante"], "Agendada": r["data_agendada"],
                "Responsável": r["responsavel"], "Status": r["status"], "Zoom": r["link_zoom"],
            })
        df = pd.DataFrame(data)
        st.caption(f"{len(df)} consultoria(s)")
        if not df.empty:
            st.dataframe(df, use_container_width=True, hide_index=True)
            ui.botoes_export(df, "consultorias")

        # Relatório por mês
        st.markdown("##### 📅 Consultorias por mês")
        rel = db.query(
            "SELECT substr(data_agendada,1,7) AS mes, COUNT(*) AS total FROM consultorias "
            "WHERE data_agendada!='' GROUP BY mes ORDER BY mes DESC")
        if rel:
            st.dataframe(pd.DataFrame([dict(r) for r in rel]), use_container_width=True, hide_index=True)

        st.divider()
        st.markdown("##### ✏️ Atualizar status")
        ids = [r["id"] for r in rows]
        if ids:
            sel = st.selectbox("Selecione", ids,
                               format_func=lambda i: next(f"#{r['id']} · {r['aluno']} ({r['tipo']})"
                                                          for r in rows if r["id"] == i))
            r = db.query_one("SELECT * FROM consultorias WHERE id=?", (sel,))
            with st.form("edit_cons"):
                c1, c2 = st.columns(2)
                status = c1.selectbox("Status", STATUS,
                                      index=STATUS.index(r["status"]) if r["status"] in STATUS else 0)
                zoom = c2.text_input("Link Zoom", r["link_zoom"] or "")
                obs = st.text_area("Observações", r["observacoes"] or "")
                if st.form_submit_button("Salvar", type="primary"):
                    db.execute("UPDATE consultorias SET status=?,link_zoom=?,observacoes=? WHERE id=?",
                               (status, zoom, obs, sel))
                    db.log_action(user["email"], "consultoria_editada", f"#{sel} {r['aluno']}")
                    st.success("Atualizado.")
                    st.rerun()

File: web/modules/consultorias.js
/* Consultorias (Private, Black, Retenção) — solicitação, agendamento e alerta Slack.
   Toda nova consultoria dispara alerta no Slack; relatório mensal por data agendada. */
import { html } from "htm/preact";
import { useState } from "preact/hooks";
import { store } from "../lib/store.js";
import { useCollection } from "../lib/hooks.js";
import {
  Badge, Table, Modal, Text, Select, DateF, Area, FilterSelect,
  toast, ExportButtons, todayISO, DOM,
} from "../lib/ui.js";
import { alertarSlack } from "../lib/notify.js";

const nowTs = () =>new Date().toISOString().slice(0, 19).replace("T", " ");
const vazia = (u) => ({ aluno: "", tipo: DOM.cons_tipo[0], solicitante: u ? u.nome : "", data_solicitada: todayISO(), data_agendada: "", responsavel: "", status: DOM.cons_status[0], link_zoom: "", observacoes: "" });

export function View({ user }) {
  const { rows } = useCollection("consultorias", { order: "data_agendada", desc: true });
  const [novo, setNovo] = useState(null);   // objeto em criação
  const [edit, setEdit] = useState(null);   // consultoria em edição
  const [fTipo, setFTipo] = useState("");
  const [fStatus, setFStatus] = useState("");
  const [fBusca, setFBusca] = useState("");

  const filtradas = rows.filter((c) =>
    (!fTipo || c.tipo === fTipo) &&
    (!fStatus || c.status === fStatus) &&
    (!fBusca || (c.aluno || "").toLowerCase().includes(fBusca.toLowerCase())));

  const salvarNova = async () => {
    if (!novo.aluno.trim()) return toast("Informe o nome do aluno.", "err");
    const c = await store.insert("consultorias", { ...novo, criado_em: nowTs() });
    await store.logAction(user.email, "consultoria_criada", `${c.tipo} · ${c.aluno}`);
    alertarSlack(`:telephone_receiver: *Nova consultoria ${c.tipo}* — ${c.aluno} (solicitante ${c.solicitante}, resp. ${c.responsavel})`, "consultoria_nova");
    setNovo(null); toast("Consultoria criada — alerta enviado ao Slack.");
  };

  const cols = [
    { key: "id", label: "ID" },
    { key: "aluno", label: "Aluno" },
    { key: "tipo", label: "Tipo" },
    { key: "solicitante", label: "Solicitante" },
    { key: "data_agendada", label: "Agendada" },
    { key: "responsavel", label: "Responsável" },
    { key: "status", label: "Status", render: (r) =>Badge(r.status), csv: (r) =>r.status },
    { key: "link_zoom", label: "Zoom",
      render: (r) =>r.link_zoom ? html`<a href=${r.link_zoom} target="_blank" onClick=${(e) =>e.stopPropagation()}>abrir</a>` : "—",
      csv: (r) =>r.link_zoom || "" },
  ];

  // agrupamento por mês (data_agendada.slice(0,7))
  const porMes = {};
  rows.forEach((r) => { const m = (r.data_agendada || "").slice(0, 7); if (m) porMes[m] = (porMes[m] || 0) + 1; });
  const meses = Object.entries(porMes).map(([mes, total]) => ({ mes, total })).sort((a, b) => (a.mes < b.mes ? 1 : -1));

  return html`
    <h1 class="h1">Consultorias</h1>
    <p class="sub">Controle de consultorias Private, Black e Retenção — cada nova consultoria alerta no Slack.</p>

    <div class="toolbar">
      <button class="btn primary" onClick=${() =>setNovo(vazia(user))}>Nova consultoria</button>
      ${FilterSelect({ label: "Tipo", value: fTipo, onInput: setFTipo, options: DOM.cons_tipo })}
      ${FilterSelect({ label: "Status", value: fStatus, onInput: setFStatus, options: DOM.cons_status })}
      <div class="grow"><label>Aluno</label>
        <input placeholder="buscar por aluno" value=${fBusca} onInput=${(e) =>setFBusca(e.target.value)}/></div>
      ${ExportButtons({ nome: "consultorias", columns: cols, rows: filtradas, titulo: "Consultorias" })}
    </div>
    <div class="count">${filtradas.length} consultoria(s)</div>
    ${Table({ columns: cols, rows: filtradas, onRow: (r) =>setEdit({ ...r }) })}

    <div class="section-t">Consultorias por mês</div>
    ${meses.length ? html`<div class="table-wrap"><table>
      <thead><tr><th>Mês</th><th>Total</th></tr></thead>
      <tbody>${meses.map((m) =>html`<tr><td>${m.mes}</td><td>${m.total}</td></tr>`)}</tbody>
    </table></div>` : html`<div class="empty">Nenhuma consultoria agendada.</div>`}

    ${novo ? html`<${Modal} title="Nova consultoria" onClose=${() =>setNovo(null)}
      footer=${html`<button class="btn" onClick=${() =>setNovo(null)}>Cancelar</button>
        <button class="btn primary" onClick=${salvarNova}>Criar consultoria</button>`}>
      ${Text({ label: "Nome do aluno *", value: novo.aluno, onInput: (v) =>setNovo({ ...novo, aluno: v }) })}
      <div class="row c3">
        ${Select({ label: "Tipo", value: novo.tipo, onInput: (v) =>setNovo({ ...novo, tipo: v }), options: DOM.cons_tipo })}
        ${Text({ label: "Solicitante", value: novo.solicitante, onInput: (v) =>setNovo({ ...novo, solicitante: v }) })}
        ${Text({ label: "Responsável interno", value: novo.responsavel, onInput: (v) =>setNovo({ ...novo, responsavel: v }) })}
      </div>
      <div class="row c3">
        ${DateF({ label: "Data solicitada", value: novo.data_solicitada, onInput: (v) =>setNovo({ ...novo, data_solicitada: v }) })}
        ${DateF({ label: "Data agendada", value: novo.data_agendada, onInput: (v) =>setNovo({ ...novo, data_agendada: v }) })}
        ${Select({ label: "Status", value: novo.status, onInput: (v) =>setNovo({ ...novo, status: v }), options: DOM.cons_status })}
      </div>
      ${Text({ label: "Link Zoom", value: novo.link_zoom, onInput: (v) =>setNovo({ ...novo, link_zoom: v }) })}
      ${Area({ label: "Observações", value: novo.observacoes, onInput: (v) =>setNovo({ ...novo, observacoes: v }) })}
    <//>` : ""}

    ${edit ? html`<${EditModal} user=${user} consultoria=${edit} onClose=${() =>setEdit(null)}/>` : ""}`;
}

function EditModal({ user, consultoria, onClose }) {
  const [c, setC] = useState({ ...consultoria });

  const salvar = async () => {
    await store.update("consultorias", consultoria.id, {
      status: c.status, link_zoom: c.link_zoom, observacoes: c.observacoes,
    });
    await store.logAction(user.email, "consultoria_editada", `#${consultoria.id} ${consultoria.aluno}`);
    toast("Consultoria atualizada."); onClose();
  };
  const excluir = async () => {
    if (!confirm("Excluir esta consultoria?")) return;
    await store.remove("consultorias", consultoria.id);
    await store.logAction(user.email, "consultoria_excluida", `#${consultoria.id}`);
    toast("Consultoria excluída."); onClose();
  };

  return html`<${Modal} title=${"Consultoria #" + consultoria.id + " · " + consultoria.aluno} size="lg" onClose=${onClose}
    footer=${html`<button class="btn danger" onClick=${excluir}>Excluir</button>
      <button class="btn" onClick=${onClose}>Fechar</button>
      <button class="btn primary" onClick=${salvar}>Salvar alterações</button>`}>
    <p class="sub">${c.tipo} · solicitante ${c.solicitante || "—"} · resp. ${c.responsavel || "—"} · agendada ${c.data_agendada || "—"}</p>
    ${Select({ label: "Status", value: c.status, onInput: (v) =>setC({ ...c, status: v }), options: DOM.cons_status })}
    ${Text({ label: "Link Zoom", value: c.link_zoom, onInput: (v) =>setC({ ...c, link_zoom: v }) })}
    ${Area({ label: "Observações", value: c.observacoes, onInput: (v) =>setC({ ...c, observacoes: v }) })}
  <//>`;
}
